### PR TITLE
Add support for CCS console access for groups

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: Ben Ye <yb532204897@gmail.com>

### What type of PR is this? 
feature & refactory

### What this PR does / why we need it:

1. As now both CCS and non-CCS needs to get the operator configmap, I move this step to the beginning of the reconcile and keep the config in the adapter struct.

2. Refactor the configmap format to yaml in order to better support a list of groups that need to get ccs console access. Now the configmap format becomes like

``` yaml
apiVersion: v1
data:
  data: |
    billingAccount: xxx
    parentFolderID: xxx
    ccsConsoleAccess:
    - example-group@xxx.com
kind: ConfigMap
metadata:
  name: gcp-project-operator
  namespace: gcp-project-operator
```

3. Unit tests updated

### Which issue(s) this PR fixes(can be a reference to existing issue or jira/bz):

Fixes https://issues.redhat.com/browse/OSD-5041 https://issues.redhat.com/browse/OSD-5046

### Pre-checks:
- [ ] manually tested latest changes against crc/k8s
- [x] run the `make coverage` command to generate new calculated coverage

### CODE REVIEW

File: README.md
Comment (mwoodson) [inline README.md:153]: We don't want the actual group name anywhere.  Let's put an example in here.

File: docs/gcpconfig.md
Comment (mwoodson) [inline docs/gcpconfig.md:30]: again, let's remove the actual group name.

File: pkg/controller/projectreference/projectreference_adapter.go
Comment (mwoodson) [inline pkg/controller/projectreference/projectreference_adapter.go:72]: please see https://docs.google.com/document/d/1vqZeayNt-GUsLJslw9jTmORvUPf1WvdhRcZgZJQeCIU/edit# for the latest list of roles

File: README.md
Comment (yeya24) [reply README.md:153]: Sure, I will update it.

File: pkg/configmap/configmap_test.go
Comment (mwoodson) [inline pkg/configmap/configmap_test.go:59]: dumb question, is this formatted properly?

File: pkg/controller/projectreference/projectreference_adapter.go
Comment (mwoodson) [inline pkg/controller/projectreference/projectreference_adapter.go:76]: If these aren't needed, we should delete them.

File: pkg/controller/projectreference/projectreference_adapter.go
Comment (mwoodson) [inline pkg/controller/projectreference/projectreference_adapter.go:251]: Is this message correct?  Is the operator configuring the service account?  Or is it configuring IAM to allow console access?

File: pkg/configmap/configmap_test.go
Comment (yeya24) [reply pkg/configmap/configmap_test.go:59]: Yes

File: pkg/controller/projectreference/projectreference_adapter.go
Comment (yeya24) [reply pkg/controller/projectreference/projectreference_adapter.go:76]: They are needed, but seems `roles/resourcemanager.organizationViewer` is deprecated and `roles/iam.organizationRoleAdmin` cannot be created using the `SetIAM` function. @sam-nguyen7

File: pkg/controller/projectreference/projectreference_adapter.go
Comment (yeya24) [reply pkg/controller/projectreference/projectreference_adapter.go:251]: `Configuring IAM to allow console access` is better! I will update it.

**Comment by mwoodson** [comment]
This PR looks good to me.  Would want @rafael-azevedo  and/or @NautiluX to also look

File: README.md
Comment (NautiluX) [inline README.md:139]:
```suggestion
You can create one example `ConfigMap` using the following command.
```

File: README.md
Comment (NautiluX) [inline README.md:153]:
```suggestion
    - example-group@xxx.com # A list of groups that will get access to CCS projects
```

File: docs/gcpconfig.md
Comment (NautiluX) [inline docs/gcpconfig.md:16]:
```suggestion
You can create one example `ConfigMap` using the following command.
```

File: docs/gcpconfig.md
Comment (NautiluX) [inline docs/gcpconfig.md:30]:
```suggestion
    - example-group@xxx.com # A list of groups that will get access to CCS projects
```

Comment (NautiluX) [reply docs/gcpconfig.md:30]: Also, since this is copied over, can't we reference it from the README or the other way round?

File: pkg/configmap/configmap.go
Comment (NautiluX) [inline pkg/configmap/configmap.go:42]: btw, why does this variable name start with a capital o?

File: pkg/configmap/configmap.go
Comment (NautiluX) [inline pkg/configmap/configmap.go:48]: this will try to get `configmap.Data["data"]` right? Is this correct?

File: pkg/configmap/configmap_test.go
Comment (NautiluX) [reply pkg/configmap/configmap_test.go:59]: Agree it's difficult to read, maybe use a one-line json?

File: pkg/configmap/configmap_test.go
Comment (NautiluX) [inline pkg/configmap/configmap_test.go:151]: see above. Using json may increase readability

File: pkg/controller/projectreference/projectreference_adapter.go
Comment (NautiluX) [reply pkg/controller/projectreference/projectreference_adapter.go:76]: that means this code is still WIP? Or you want to merge it like this for now but we know it won't work for future changes?

File: pkg/controller/projectreference/projectreference_adapter.go
Comment (NautiluX) [inline pkg/controller/projectreference/projectreference_adapter.go:657]: I'd prefer to not use a bool as type for group, this makes function calls hard to read. I recommend using a type for it.

**Review by NautiluX — CHANGES_REQUESTED** [review]
Hi @yeya24, thanks for submitting this PR! I made a few comments. My biggest concern is the format of the config map, currently it looks like
```
apiVersion: v1
data:
  billingAccount: asdf
  orgParentFolderID: asdf
  parentFolderID: asdf
kind: ConfigMap
metadata:
  annotations:
  name: gcp-project-operator
  namespace: gcp-project-operator
```

Are you sure the new structs are able to parse this kind of config map? Because in the example and implementation it looks like there was an additional level in them, which confuses me:

```
data:
  data: |
```

File: pkg/controller/projectreference/projectreference_adapter.go
Comment (yeya24) [reply pkg/controller/projectreference/projectreference_adapter.go:76]: This part is blocked https://coreos.slack.com/archives/CMD1L32M9/p1600780878019500. I think it is okay to ignore this first. WDYT? I can add a TODO.

File: pkg/configmap/configmap.go
Comment (yeya24) [reply pkg/configmap/configmap.go:42]: Nice catch!

File: pkg/configmap/configmap.go
Comment (yeya24) [reply pkg/configmap/configmap.go:48]: Yes, it requires the `data` key.

File: README.md
Comment (yeya24) [reply README.md:153]: Thanks! It is much better!

File: docs/gcpconfig.md
Comment (yeya24) [reply docs/gcpconfig.md:30]: Great idea. I will add a link from readme to the config doc.

File: pkg/controller/projectreference/projectreference_adapter.go
Comment (sam-nguyen7) [reply pkg/controller/projectreference/projectreference_adapter.go:76]: I will be reaching out to the google TAM to follow up on this. @yeya24 you can remove these roles for this PR. Thank you!

File: pkg/controller/projectreference/projectreference_adapter.go
Comment (yeya24) [reply pkg/controller/projectreference/projectreference_adapter.go:657]: WDYT @rafael-azevedo @sam-nguyen7 ?

File: pkg/configmap/configmap_test.go
Comment (yeya24) [reply pkg/configmap/configmap_test.go:59]: Umm, why this is a problem. That's how `` is used to express yaml. It would be clear in the real yaml config.

**Comment by yeya24** [comment]
> Are you sure the new structs are able to parse this kind of config map? Because in the example and implementation it looks like there was an additional level in them, which confuses me:

It cannot parse the old format and the config map needs to be updated.

File: pkg/configmap/configmap.go
Comment (NautiluX) [reply pkg/configmap/configmap.go:48]: Can't we find a way to parse the existing format? I like the parsing, but having
```
data:
  data:
```
seems unclean to me.

File: pkg/controller/projectreference/projectreference_adapter.go
Comment (NautiluX) [reply pkg/controller/projectreference/projectreference_adapter.go:655]: would it be a high effort to get this done here? I just fear code we add with the intention to change "later" because often enough later means never unfortunately.

Comment (yeya24) [reply pkg/controller/projectreference/projectreference_adapter.go:655]: Makes sense! But tbh I cannot figure out that API and it is not a high priority. So I will change bool to a type first.

File: pkg/configmap/configmap.go
Comment (yeya24) [reply pkg/configmap/configmap.go:48]: It is hard to still parse the existing format. Since config map itself is `map[string]string`, if we want to have a list, then we only have 2 choices:

1. Have a single key, and put all our configs to the value. If you think the key name `data` not good, I can update it to something else, like `config.yaml`, so people will know it is actually a yaml config file. WDYT?

``` yaml
data:
  config.yaml: |
    billingAccount: xxx
    folderID: xxx
    consoleAccess: [item1, item2]
```

2. Have multiple keys. In this case, we have to find a way to make consoleAccess value a string. TBH it might doable, but maybe mess up things.

``` yaml
data:
  billingAccount: xxx
  folderID: xxx
  consoleAccess: "item1, item2"
```

File: pkg/controller/projectreference/projectreference_adapter.go
Comment (yeya24) [reply pkg/controller/projectreference/projectreference_adapter.go]: Updated, sorry for the delay. Please take a look!

File: pkg/configmap/configmap.go
Comment (yeya24) [reply pkg/configmap/configmap.go:48]: I am still looking at this and I find a way to get this working for both formats.

``` yaml
data:
  billingAccount: xxx
  folderID: xxx
  config.yaml: |
    billingAccount: xxx
    folderID: xxx
    consoleAccess: [item1, item2]
```

**Comment by yeya24** [comment]
Can I get another review for this pr? @NautiluX @sam-nguyen7

File: pkg/configmap/configmap.go
Comment (NautiluX) [reply pkg/configmap/configmap.go:48]: yes @yeya24 let's do it like this and switch to the new format now, I think it's the cleaner way of defining and parsing a configmap. Some thoughts:
- Make sure it works if only the "old" fields are set
- Don't call it `data`, maybe `config.yaml` would be a good alternative.
- Let's not defer cleaning up the redundant fields but prepare a PR right away so we can clean it up as soon as all environments are migrated.

Comment (lisa) [reply pkg/configmap/configmap.go:48]: 👍 to `config.yaml` key name.

**Review by NautiluX — APPROVED** [review]
/lgtm
/approve
/hold pending deployment of new config maps

**Comment by NautiluX** [comment]
/unhold
config maps updated

**Pull request merged**
